### PR TITLE
Merge recommendation configuration with override strategy

### PR DESCRIPTION
**Pull request by pkosiec**
## Description

Changes proposed in this pull request:

- Merge recommendation configuration with override strategy

As pointed by @mszostok in https://github.com/kubeshop/botkube-docs/pull/122#discussion_r944264267, it would be good to have unification over merging strategies.

## Testing

See instruction from: https://github.com/kubeshop/botkube/pull/687
Apart from unit test, I tested the changes manually.

## Related issue(s)

https://github.com/kubeshop/botkube/issues/684

### CODE REVIEW

File: pkg/config/config.go [hunk 1]
```diff
@@ -181,19 +181,19 @@ type Recommendations struct {
 // PodRecommendations contains configuration for pods recommendations.
 type PodRecommendations struct {
 	// NoLatestImageTag notifies about Pod containers that use `latest` tag for images.
-	NoLatestImageTag bool `yaml:"noLatestImageTag"`
```
Comment (mszostok) [inline pkg/config/config.go:184]: for all pointers I would add `,omitempty` tag

File: pkg/recommendation/factory.go [hunk 2]
```diff
@@ -34,36 +34,46 @@ func NewFactory(logger logrus.FieldLogger, dynamicCli dynamic.Interface) *Factor
 }
 
 // NewForSources merges recommendation options from multiple sources, and creates a new Set.
-func (f *Factory) NewForSources(sources map[string]config.Sources) Set {
+func (f *Factory) NewForSources(sources map[string]config.Sources, mapKeyOrder []string) Set {
```
Comment (mszostok) [inline pkg/recommendation/factory.go:37]: IMO it can be simplified a bit:
```go

// NewForSources merges recommendation options from multiple sources, and creates a new Set.
func (f *Factory) NewForSources(sources map[string]config.Sources, mapKeyOrder []string) Set {
	
	rec := config.Recommendations{}
	for _, key := range mapKeyOrder {
		source, exists := sources[key]
		if !exists {
			continue
		}

		sourceRec := source.Kubernetes.Recommendations
		if sourceRec.Pod.LabelsSet != nil {
			rec.Pod.LabelsSet = sourceRec.Pod.LabelsSet
		}
		if sourceRec.Pod.NoLatestImageTag != nil {
			rec.Pod.NoLatestImageTag = sourceRec.Pod.NoLatestImageTag
		}
		if sourceRec.Ingress.BackendServiceValid != nil {
			rec.Ingress.BackendServiceValid = sourceRec.Ingress.BackendServiceValid
		}
		if sourceRec.Ingress.TLSSecretValid != nil {
			rec.Ingress.TLSSecretValid = sourceRec.Ingress.TLSSecretValid
		}
	}

	set := f.enableRecommendationIfShould(rec)
	return newRecommendationsSet(f.logger, set)
}


func (f *Factory) enableRecommendationIfShould(rec config.Recommendations) map[string]Recommendation {
	var list []Recommendation
	if isTrue(rec.Pod.LabelsSet) {
		list = append(list, NewPodLabelsSet())
	}

	if isTrue(rec.Pod.NoLatestImageTag) {
		list = append(list, NewPodNoLatestImageTag())

	}

	if isTrue(rec.Ingress.BackendServiceValid) {
		list = append(list, NewIngressBackendServiceValid(f.dynamicCli))

	}

	if isTrue(rec.Ingress.TLSSecretValid) {
		list = append(list, NewIngressTLSSecretValid(f.dynamicCli))
	}

	out := map[string]Recommendation{}
	for _, rec := range list {
		out[rec.Name()] = rec
	}
	return out
}


func isTrue(in *bool) bool {
	if in == nil {
		return false
	}
	return *in
}
```

Currently, I see two problems:
- enableRecommendationIfShould - it's not only about "enable" or not it's about construct/destruct obj
- in a loop we create and destruct obj, one thing is GC but I don't care about that too much right now, for me it was just hard to follow the logic of it.

File: pkg/recommendation/factory.go [hunk 3]
```diff
@@ -34,36 +34,46 @@ func NewFactory(logger logrus.FieldLogger, dynamicCli dynamic.Interface) *Factor
...
-		f.createIfNotExist(k8sSource.Recommendations.Ingress.BackendServiceValid, set,
+		f.enableRecommendationIfShould(k8sSource.Recommendations.Ingress.BackendServiceValid, set,
 			func() Recommendation { return NewIngressBackendServiceValid(f.dynamicCli) },
 		)
-		f.createIfNotExist(k8sSource.Recommendations.Ingress.TLSSecretValid, set,
+		f.enableRecommendationIfShould(k8sSource.Recommendations.Ingress.TLSSecretValid, set,
 			func() Recommendation { return NewIngressTLSSecretValid(f.dynamicCli) },
 		)
 	}
 	return newRecommendationsSet(f.logger, set)
```
Comment (mszostok) [inline pkg/recommendation/factory.go:59]: do we need to really have a set? what about `[]Recommendation` so that would be the `enableRecommendationIfShould` output:
```go
func (f *Factory) enableRecommendationIfShould(rec config.Recommendations) []Recommendation {
	var list []Recommendation
	if isTrue(rec.Pod.LabelsSet) {
		list = append(list, NewPodLabelsSet())
	}

	if isTrue(rec.Pod.NoLatestImageTag) {
		list = append(list, NewPodNoLatestImageTag())

	}

	if isTrue(rec.Ingress.BackendServiceValid) {
		list = append(list, NewIngressBackendServiceValid(f.dynamicCli))

	}

	if isTrue(rec.Ingress.TLSSecretValid) {
		list = append(list, NewIngressTLSSecretValid(f.dynamicCli))
	}

	return list
}
```

Comment (pkosiec) [reply pkg/recommendation/factory.go:59]: Sure, if we already ensured there are single instances of each validator, then yes, it can be a slice 👍

**Review by mszostok — APPROVED** [review]
